src/RateLimiter/CacheCounter.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Web\RateLimiter;
+
+use Psr\SimpleCache\CacheInterface;
+
+/**
+ * CacheCounter implements generic сell rate limit algorithm https://en.wikipedia.org/wiki/Generic_cell_rate_algorithm
+ */
+final class CacheCounter
+{
+    public const ID_PREFIX = 'rate-limiter-';
+
+    private const MILLISECONDS_PER_SECOND = 1000;
+
+    private int $period;
+
+    private int $limit;
+
+    private ?string $id = null;
+
+    private CacheInterface $storage;
+
+    private int $arrivalTime;
+
+    public function __construct(int $limit, int $period, CacheInterface $storage)
+    {
+        if ($limit < 1) {
+            throw new \InvalidArgumentException('The limit must be a positive value.');
+        }
+
+        if ($period < 1) {
+            throw new \InvalidArgumentException('The period must be a positive value.');
+        }
+
+        $this->limit = $limit;
+        $this->period = $period * self::MILLISECONDS_PER_SECOND;
+        $this->storage = $storage;
+    }
+
+    public function setId(string $id): void
+    {
+        $this->id = self::ID_PREFIX . $id;
+    }
+
+    public function incrementAndGetResult(): RateLimitResult
+    {
+        if ($this->id === null) {
+            throw new \LogicException('The counter id not set');
```

Comment:
It looks like setting `id` is mandatory. If so, it worth moving it to constructor.

Reply:
fixed